BUG: ndarrays pickled by 1.16 cannot be loaded by 1.15.4 and lower

mattip opened the pull request:
Fixes #12837 
When pickling ndarrays, the pickle protocol adds the reconstructing function's __module__` name to the pickled string so it can be called. The function (which is `_reconstruct`) now lives in `numpy.core._multiarray_umath`, but in pre-1.16 numpy it was in `numpy.core.multiarray`. Also for scalar objects, the function `scalar` moved.

The fix is very simple, changing the `__module__` attribute of `_reconstruct` and `scalar`

In `array_reduce_ex`, used when the pickle protocol is 5, the function will be `numpy.core.numeric._frombuffer`, which will not exist on pre-1.16 numpy (added in commit 64a855f421, Oct 2018). I don't think we can work around that. People who do not have that function will have to modify their numpy if they wish to import protocol-5 pickled ndarrays.

mattip committed:
BUG: ndarrays pickled by 1.16 cannot be loaded by 1.15.4 and lower

charris commented on numpy/core/tests/test_regression.py:
```diff
@@ -2411,6 +2411,79 @@ def test_field_access_by_title(self):
+        # Use \n to allow pretty formatting here but no SyntaxError in Popen
+        # due to lines beginning with spaces
+        cmd = [new_python, '-s', '-c', f'''
+            \n# -------Make sure _multiarray_umath cannot be imported --------
```
I don't think you need the `\n` here, line breaks are preserved with the triple quotes.

mattip replied:
If I don't use `\n` I need to left-justify the text in the file, so the block would not be nicely indented. Would that be OK?

mattip replied:
used dedent

charris commented:
Lots of test failures.

charris commented:
Cannot backport if it uses fstrings.

mattip commented:
Removed fstrings, rebased off master, but the test still causes problems on travis. It seems creating a venv via `pytohn3 -venv` and installing things into it somehow messes up the current environment on travis, azure seems OK with it.

I am not sure why the `numpy/distutils/tests/test_exec_command.py` test is failing, see PR #12849

mattip commented on numpy/core/tests/test_regression.py:
```diff
@@ -2411,6 +2411,80 @@ def test_field_access_by_title(self):
+    @pytest.mark.skipif(not sys.platform.startswith('linux'),
+                        reason="linux only")
+    @pytest.mark.skipif(sys.version_info < (3,4), reason="lacks venv")
+    @pytest.mark.skip("venv on travis is flakey")
```
Skip the test entirely for now, maybe someday re-enable. It passes when run alone (`--k test_pickle_across_versions`) which gives me confidence that the actual fix works

charris replied:
How about just look at the pickle header?

charris replied:
Hmm, OK, looks like pickle itself determines the module, but it must be in the header somewhere. I wonder if the pickle can simply be searched for the module name? For instance
```
In [30]: a = array(1)

In [31]: pickle.dumps(a, protocol=4)
Out[31]: b'\x80\x04\x95\x96\x00\x00\x00\x00\x00\x00\x00\x8c\x1cnumpy.core._multiarray_umath\x94\x8c\x0c_reconstruct\x94\x93\x94\x8c\x05numpy\x94\x8c\x07ndarray\x94\x93\x94K\x00\x85\x94C\x01b\x94\x87\x94R\x94(K\x01)h\x03\x8c\x05dtype\x94\x93\x94\x8c\x02i8\x94K\x00K\x01\x87\x94R\x94(K\x03\x8c\x01<\x94NNNJ\xff\xff\xff\xffJ\xff\xff\xff\xffK\x00t\x94b\x89C\x08\x01\x00\x00\x00\x00\x00\x00\x00\x94t\x94b.'
```

charris replied:
Looking at that, I wonder why the umath module is there? Maybe from the ndarray `__reduce__` method? AFAICT that method should not be called, being replaced by `__reduce_ex__`. I think we should look at both those methods.

charris replied:
We should also test the different protocols.

charris replied:
If you really want to test across versions, could also pickle some ndarray in an earlier version, get the hash of the result, and check that it agrees with the hash of the same data pickled with a latter version. The "correct" hashes can be stored in the test.

charris replied:
For instance with current master and previous array
```
In [43]: hashlib.sha256(s).hexdigest()
Out[43]: 'cce278984d68606f0ceb4d23c298ebd8abe944adba1bc6b161f16c180e785fc3'
```

charris replied:
The `array_reduce` function does look buggy, but it is unlikely it gets called unless some ndarray subclass does it explicitly.

mattip replied:
> I wonder why the umath module is there

The module is `_multiarray_umath`. 

I will write a test that checks the bytestring for the correct module name in all protocols. Tieing ourselves to the exact pickle bytestring or hash seems too restrictive going forward.

charris commented:
There is also some ufunc pickling stuff in `numpy/core/__init__.py` that might could have problems.

charris commented:
The ufunc pickles look OK.

charris commented:
Is `_reconstruct` the only problem?

mattip commented:
`_reconstruct` for ndarrays and `scalar` for scalars both need their `__module__` adjusted

charris commented on azure-pipelines.yml:
```diff
@@ -14,7 +14,7 @@ jobs:
            docker run -v $(pwd):/numpy i386/ubuntu:bionic /bin/bash -c "cd numpy && \
            apt-get -y update && \
-           apt-get -y install python3.6-dev python3-pip locales && \
+           apt-get -y install python3.6-dev python3-venv python3-pip locales && \
```
No longer need `python3-venv`?

mattip replied:
True, but on the other hand it couldn't hurt and makes this environment consistent with the other linux ones one travis

charris replied:
Come, come, you could install lots of things that don't hurt, but that is not a reason to do it.

mattip replied:
ok, removing

charris commented:
LGTM, one possible nit.

charris commented:
Thanks Matti.
